### PR TITLE
SDIT-902 Refactor tests

### PR DESCRIPTION
**Pull request by mikehalmamoj**

### CODE REVIEW

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/ProgramService.kt [hunk 1]
```diff
@@ -9,11 +9,33 @@ import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.repository.ProgramServi
 annotation class ProgramServiceDslMarker
 
 @NomisDataDslMarker
-interface ProgramServiceDsl : CourseActivityDslApi
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/ProgramService.kt:12]: Removed all DSL API interfaces and moved the abstract methods inline. This aligns with the DSL pattern used elsewhere.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/ProgramService.kt [hunk 2]
```diff
@@ -12,11 +14,13 @@ import org.hibernate.type.YesNoConverter
 @Table(name = "PROGRAM_SERVICES")
 data class ProgramService(
   @Id
+  @SequenceGenerator(name = "PROGRAM_ID", sequenceName = "PROGRAM_ID", allocationSize = 1)
+  @GeneratedValue(generator = "PROGRAM_ID")
   @Column(nullable = false)
   val programId: Long = 0,
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/ProgramService.kt:20]: Use a sequence for the program ID like Nomis does.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/ProgramService.kt [hunk 3]
```diff
@@ -9,11 +9,33 @@ import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.repository.ProgramServi
...
+interface ProgramServiceDsl {
+  @CourseActivityDslMarker
+  fun courseActivity(
+    courseActivityId: Long = 0,
+    code: String = "CA",
+    prisonId: String = "BXI",
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/ProgramService.kt:17]: Changed default prisons to BXI as this is also the default prison on an offender. This means we don't have to include the prison ID in test data setup where it is largely irrelevant wrt what is being tested.

File: src/main/resources/db/migration/data/R__3_2__AGENCY_INTERNAL_LOCATIONS.sql [hunk 4]
```diff
@@ -331,4 +331,6 @@ INSERT INTO AGENCY_INTERNAL_LOCATIONS (INTERNAL_LOCATION_ID, INTERNAL_LOCATION_C
...
 INSERT INTO AGENCY_INTERNAL_LOCATIONS (INTERNAL_LOCATION_ID, INTERNAL_LOCATION_CODE, INTERNAL_LOCATION_TYPE, AGY_LOC_ID, UNIT_TYPE, CERTIFIED_FLAG, CAPACITY, OPERATION_CAPACITY, DESCRIPTION, PARENT_INTERNAL_LOCATION_ID, USER_DESC, ACTIVE_FLAG, DEACTIVATE_DATE) VALUES (-3004, 'VISIT2', 'VISIT', 'BXI', null, 'N', 10, null, 'BXI-VISIT2', null, 'B room 1', 'Y', NULL);
+INSERT INTO AGENCY_INTERNAL_LOCATIONS (INTERNAL_LOCATION_ID, INTERNAL_LOCATION_CODE, INTERNAL_LOCATION_TYPE, AGY_LOC_ID, UNIT_TYPE, CERTIFIED_FLAG, CAPACITY, OPERATION_CAPACITY, DESCRIPTION, PARENT_INTERNAL_LOCATION_ID, USER_DESC, ACTIVE_FLAG, DEACTIVATE_DATE) VALUES (-3005, 'CLASS1', 'CLAS',  'BXI', 'NA', 'Y', 10, null, 'BXI-CLASS1', null, 'Classroom 1', 'Y', NULL);
+INSERT INTO AGENCY_INTERNAL_LOCATIONS (INTERNAL_LOCATION_ID, INTERNAL_LOCATION_CODE, INTERNAL_LOCATION_TYPE, AGY_LOC_ID, UNIT_TYPE, CERTIFIED_FLAG, CAPACITY, OPERATION_CAPACITY, DESCRIPTION, PARENT_INTERNAL_LOCATION_ID, USER_DESC, ACTIVE_FLAG, DEACTIVATE_DATE) VALUES (-3006, 'CLASS2', 'CLAS',  'BXI', 'NA', 'Y', 10, null, 'BXI-CLASS2', null, 'Classroom 2', 'Y', NULL);
```
Comment (mikehalmamoj) [inline src/main/resources/db/migration/data/R__3_2__AGENCY_INTERNAL_LOCATIONS.sql:335]: Create a couple of dedicated classrooms in BXI to use in tests.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ActivityResourceIntTest.kt [hunk 5]
```diff
@@ -718,8 +745,8 @@ class ActivityResourceIntTest : IntegrationTestBase() {
               courseScheduleRule()
             }
           }
-          offender(nomsId = "A1234TT") {
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ActivityResourceIntTest.kt:721]: Use the default nomsId wherever possible, it's largely irrelevant to what is being tested and is just noise.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ActivityResourceIntTest.kt [hunk 6]
```diff
@@ -100,7 +100,7 @@ class ActivityResourceIntTest : IntegrationTestBase() {
 
       @Test
       fun `invalid prison should return bad request`() {
-        val invalidPrison = validJsonRequest().replace(""""prisonId" : "LEI",""", """"prisonId" : "ZZX",""")
+        val invalidPrison = validJsonRequest().withPrison("ZZX")
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ActivityResourceIntTest.kt:103]: Lots of these to standardise JSON request creation already in use in some test classes - hide the details of JSON wrangling for tests to make the tests more readable.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/AttendanceResourceIntTest.kt [hunk 7]
```diff
@@ -502,17 +502,6 @@ class AttendanceResourceIntTest : IntegrationTestBase() {
...
-        nomisDataBuilder.build {
-          offender(nomsId = "A1234TU") {
-            offenderBooking = booking(agencyLocationId = "MDI") {
-              allocation = courseAllocation(courseActivity) {
-                payBand()
-                attendance = courseAttendance(courseSchedule)
-              }
-            }
-          }
-        }
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/AttendanceResourceIntTest.kt:514]: Identical to default test setup in `@BeforeEach` so not needed.

**Review by lbennett-moj — APPROVED** [review]

**Pull request merged**